### DISCUSSION TITLE
Can we useQuasar in a composable?

### DISCUSSION THREAD

**Discussion by bvbellomo** [quasarframework/quasar#12194] [General - Components / Directives / etc] [answered]
If I put this code inside a component, it works:

const $q = useQuasar();
$q.dialog({title: 'Alert', message: 'Some message' });

If I try calling from a composable, or any function defined inside a ts file that exports a composable, it does not work.
It appears $q is undefined.

I also see "[Vue warn]: inject() can only be used inside setup() or functional components" although that might be unrelated.

Is this by design?  Is there a way to create a dialog in a function in a composable?

**Comment by yusufkandemir** [comment, marked as answer]
As long as the call to `useQuasar()`, `useRouter()`, `useStore()`, or any other composable that uses `inject` under the hood, is in the same call stack as the composable you call first inside `setup()`, it should work. You might be calling it outside the function at the top level of the file, or maybe inside Vuex, or anywhere else that is not inside `setup()`.

Outside the `setup()` scope, in boot files, etc. you need to directly import the plugin and use it directly. You can still use this approach without drawbacks unless the plugin states it explicitly in its documentation. An example of this would be `Cookies` plugin regarding SSR.

You can import and use the Dialog plugin like this:
```js
import { Dialog } from 'quasar';
// ...
Dialog.create({ title: 'Alert', message: 'Some message' });
```

This approach will work, but there might be something wrong in the way you were calling `useQuasar`, so I would suggest understanding the whole mechanism first. Because you might face situations where you don't have any alternatives.
